Add in basic support for OOM retry for project and filter

revans2 opened the pull request:
This adds in basic support for retry for project and filter. It specifically adds that retry in for hash aggregate pre and post processing. The pre-processing can be quite memory intensive and in our testing is likely to be one of the places where GPU OOM errors often happen.

This is not 100% ready to go in. I am working on some unit tests for the code. I have manually tested it, but I would like some more automated tests.

There is still a lot of follow on work to do too.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:
```diff
@@ -332,34 +408,14 @@ case class GpuProjectAstExec(
         }
       case exprSet :: tail =>
         val projectCb = withResource(new NvtxRange("project tier", NvtxColor.ORANGE)) { _ =>
```
it seems we should `closeOnExcept(cb)` while doing the project

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:
```diff
@@ -283,46 +344,61 @@ case class GpuProjectAstExec(
+              spillCallback)
+          }
+          // Close intermediate batches
+          if (!isFirst) sb.close()
```
I think this is double closing `sb`, but it is ok because RapidsBufferHandle protects against that, that said we shouldn't need this line.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:
```diff
@@ -283,46 +344,61 @@ case class GpuProjectAstExec(
+  def projectAndCloseWithRetrySingleBatch(sb: SpillableColumnarBatch,
+      spillCallback: SpillCallback): ColumnarBatch = {
+    if (areAllDeterministic) {
+      // If all of the expressions are deterministic we can just run everything are retry it
```
nit,
```suggestion
      // If all of the expressions are deterministic we can just run everything and retry it
```

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:
```diff
@@ -116,6 +110,71 @@ object GpuProjectExec extends Arm {
+  /**
+   * Just like project, but it will try and retry the operations if it can.
+   * @param sb the input batch
+   * @param boundExprs the expressions to run
```
nit, missing `spillCallback`

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:
```diff
@@ -116,6 +110,71 @@ object GpuProjectExec extends Arm {
+   * @param boundExprs the expressions to run
+   * @return the resulting batch
+   */
+  def projectAndCloseWithRetrySingleBatch(sb: SpillableColumnarBatch,
```
might be nice to clarify this is internal (to be used for a specific tier) as far as I understand.

revans2 replied:
It isn't totally that way.  It can be used stand alone, but the Tiered API I think is better/simpler to use.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:
```diff
@@ -116,6 +110,71 @@ object GpuProjectExec extends Arm {
+    // These cannot be retried.
+    val nonDeterministicExprs = ArrayBuffer[Expression]()
+    val deterministicExprs = ArrayBuffer[Expression]()
+    boundExprs.foreach { expr =>
```
I think `val (deterministicExprs, nonDeterministicExprs) = boundExprs.partition(_.deterministic)` would do also.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:
```diff
@@ -116,6 +110,71 @@ object GpuProjectExec extends Arm {
+      // First run the nonDeterministic ops because we cannot retry them
+      // In the future we might be able to isolate the non-deterministic expressions and do it more
+      // of a tiered operation
+      if (nonDeterministicExprs.nonEmpty) {
```
nit, maybe nice to move this check before we do `sb.getColumnarBatch()`, so we don't need to bring back something from a spilled buffer potentially back to just close it?

revans2 commented:
@abellina I added a few tests and addressed your review feedback. I think I am done at this point. I cannot easily test retry on non-deterministic project. It is a rare enough case that I decided not to test it outside of normal integration tests, but if you want me to I am happy to try and add something in.

revans2 commented:
vulnerability scan is failing with a docker error where a checksum is not registered. Not sure how to work around this...

revans2 commented:
build

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:
```diff
@@ -332,34 +408,14 @@ case class GpuProjectAstExec(
-              GpuColumnVector.combineColumns(remainingCb, newCols)
-            }
-          }
+          GpuProjectExec.project(cb, exprSet)
```
do we not need to protect against `project` throwing for intermediate batches?

abellina replied:
never mind, you did that :)

abellina commented:
Overall I just had a question on the temporary batches as we recurse, the rest looks good to me. It would be great to get @jbrennan333's feedback though.

abellina reviewed (approved):
LGTM

revans2 commented:
I ran some performance tests locally. I need to do it in a more stable environment, but it looks to be around a 3% performance regression on NDS with this patch. I think that is acceptable especially if we have a plan on how to improve the likely culprit (extra contig split calls to make the data spillable).

jbrennan333 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBoundAttribute.scala:
```diff
@@ -130,19 +130,41 @@ object GpuBindReferences extends Logging {
-    }, inputTiers)
+      input: AttributeSeq,
+      runTiered: Boolean): GpuTieredProject = {
+
```
I wish I had thought to move pass runTiered like this.  Definitely cleans up the call sites.

jbrennan333 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBoundAttribute.scala:
```diff
@@ -130,19 +130,41 @@ object GpuBindReferences extends Logging {
+      val exprTiers = GpuEquivalentExpressions.getExprTiers(expressions)
+      val inputTiers = GpuEquivalentExpressions.getInputTiers(exprTiers, input)
+      // Update ExprTiers to include the columns that are pass through and drop unneeded columns
+      val newExprTiers = exprTiers.zipWithIndex.map {
+        case (exprTier, index) =>
+          // get what the output should look like.
+          val atInput = index + 1
+          if (atInput < inputTiers.length) {
+            inputTiers(atInput).attrs.map { attr =>
+              exprTier.find { expr =>
+                expr.asInstanceOf[NamedExpression].toAttribute == attr
+              }.getOrElse(attr)
+            }
+          } else {
+            exprTier
+          }
+      }
```
Really nice refactoring of the code to remove the unneeded columns - simplifies the project code.

jbrennan333 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:
```diff
@@ -283,83 +336,72 @@ case class GpuProjectAstExec(
+      }
+    } else {
+      @tailrec
+      def recurse(boundExprs: Seq[Seq[GpuExpression]],
```
(nit) maybe stylistic, but I'd prefer if this recurse inner function were moved up to the top of this function.

jbrennan333 reviewed (approved):
lgtm.  Just one minor nit.

revans2 commented:
build
